### ISSUE TITLE
Display driver crashes

### ISSUE THREAD

**Issue by merceyz** [jcredi/trafficNetCL#1]
My display driver crashes quite often when using this. 

I can use it for a while then the screen starts displaying black boxes at random places (running under 60*C so it's not overheating) and eventually crash.

Sometimes when starting it gets to evaluate and the display driver crashes (and recovers) and it just sits there

GPU: MSI R9 280X 3GB Gaming @ 1020 MHz Core, 1500 MHz Memory

OpenCL output: 

> Platform: AMD Accelerated Parallel Processing
> Device 0: Tahiti
> Device 1: Intel<R> Core<TM> i7-2600 CPU @ 2.40GHz
> 
> Using Device Tahiti
> Max work item sizes: 256, 0, 256
> Max work group size: 256
> Max compute units: 32

Please let me know if more data is needed

**Comment by jcredi** [comment]
Hi merceyz,
Thanks for trying out the Conv.NET library!

I'm sorry to hear you're having trouble using it. Unfortunately I have only tested it on Nvidia and Intel hardware, and currently I don't have an AMD GPU to test it on. I will try to get my hands on one and come back to you as soon as possible.

In the meantime, can you tell me more about what kind of network you are creating (layers, etc.) and the size of your dataset?

**Comment by merceyz** [comment]
If you want to I can run tests on my system and report back the results, so you don't have to go out and buy a card just for me

I'm trying to have it learn 96x96x3 images like these (water):
![6c153eca6c5d8152b2bb8daf5dd380e6](https://cloud.githubusercontent.com/assets/3842800/16850921/79e099ba-4a01-11e6-8111-ea5bf76354da.png)

There is a total of 3 labels, their example (image) count is as follows:

> 0: 401
> 1: 53
> 2: 374

Since the DataSet class didn't have a option to supply it with bitmaps I created my own which looks like this:
![d772b266c741f069b3e7edf99b01ffa1](https://cloud.githubusercontent.com/assets/3842800/16851040/06b54a34-4a02-11e6-9f1b-10a36eaec0a5.png)
In my usage the dataPoint has a dimension of 96x96x3 = 27648
First 9216 floats are the R value, 9216 to 18432 are the G value and 18432 to 27648 is the B value
R, G and B are values between 0 and 255

Network is created using this code:
![c322c6dcfdd47a87de0c6a7c8425082a](https://cloud.githubusercontent.com/assets/3842800/16852471/9371ace6-4a08-11e6-926f-4076b229ea83.png)

NetworkTrainer's values:
![a59c988fd4c75fcc51c06c978a8106f1](https://cloud.githubusercontent.com/assets/3842800/16851279/1f4e6214-4a03-11e6-8407-62e209d41dbc.png)

Some errors i get every now and then:
- > Exception thrown at 0x000007FEDA9AE6D4 (amdocl64.dll) in Console for testing.exe: 0xC0000005: Access violation reading location 0x0000000000000019.
  > 
  > If there is a handler for this exception, the program may be safely continued.
- > Unhandled exception at 0x00000000777FF262 (ntdll.dll) in Console for testing.exe: 0xC0000374: A heap has been corrupted (parameters: 0x0000000077869430).

I also noticed that if i give it an image i wont get the same output if i run it through again, how come?

If there is anything else you need, please let me know

**Comment by jcredi** [comment]
Thanks for providing extensive info, it's very appreciated. 

First of all, my apologies for the poor quality of my code and even poorer documentation. I planned to clean it up and improve documentation in my free time, but in all honesty I did not expect someone other than me to actually use my code, or at least not so soon.

The errors you get reminded me of an issue I encountered a couple of months ago. I remember I tried out the AMD OpenCL compiler in the AMD APP SDK on my Intel CPU (i7-4700MQ) and it turned out to be slightly faster than the Intel OpenCL compiler. It was also more unstable, though, and I got similar errors, in particular when using pooling layers. However such errors disappeared when using Nvidia OpenCL platform, and I then used both a GTX 850M and a GTX 980Ti without any issue. I also clearly remember I found an official AMD APP SDK bug report about the issue I was having, so I didn't bother digging too deep into it.

Now, you can try the following:
- Download the [OpenCL drivers for Intel Architecture](https://software.intel.com/en-us/articles/opencl-drivers) and try using your CPU instead. If you don't get exceptions, then my code may indeed be incompatible with the AMD APP platform. If you do encounter problems, then my code is likely buggy (or, well, it's even _more_ buggy than I thought), in which case I'll try to locate the bug.
- Remove MaxPooling layers and see if you still get those exceptions. If the computational burden becomes too large, you can use a stride of 2 in the 2nd, 4th, and 6th convolutional layers to decrease spatial dimensions. 

The ReadBitmap() method seems to do the job well. That's exactly how the data should be stored in the DataSet class. However, in my opinion your dataset may be a bit too small to train a 12-layer net from scratch. Have you considered using a pretrained model, e.g. on ImageNet? There are several really good models out there, and you might consider using pretrained conv-layers and only train a classifier on top of them (fully-connected layers).

Regarding your last question, I assume you are passing the same image twice using the NetworkEvaluator, or using the NetworkTrainer with a learning rate of zero, right? (If not, then obviously you _should_ get a different output, as the network parameters are updated after each mini-batch!) If that's the case, it's very weird. I'll run some tests and let you know as soon as I can.

**Comment by merceyz** [comment]
You're very welcome, as a developer I know how it's like to have people go "it doesn't work" and never hear from them again.

No need to excuse yourself, you've done a mighty fine job. You have no idea how many libraries/frameworks I've tried, all of which were terrible or didn't even compile.

I downloaded the OpenCL drivers for my CPU which resulted in the OpenCL output to change from 2 devices to 3
![a8b8640cc8cd77a19f201b738b852b4f](https://cloud.githubusercontent.com/assets/3842800/16878043/44696b5e-4aab-11e6-8584-a54277a64ad2.png)
I used device number 3(2) and ran the exact same code with the same network layout.
It doesn't crash but is incredibly slow (compared to GPU), but it doesn't crash and no black boxes.

I switched back to my GPU and did as you suggested, which makes the creation code look as follows
![05fe19c76af8bde3152b7b697b65bd72](https://cloud.githubusercontent.com/assets/3842800/16878376/9a7d707a-4aac-11e6-9de4-572ba633fdb3.png)
Now it didn't throw an exception nor crash the driver but i still got some black boxes every now and then.
On a positive side it now uses ~38000ms per epoch compared to the ~108000ms it used with the maxpooling and the loss and error values both went down to around 0,4 and 0,1.

EDIT: It just crashed now when I moved the VS window with the program not running

I'm glad my method seems to work well (copy paste implementations are the best right?) as for the DataSet size it is growing constantly with the amount of examples and for the network I have to be honest. I kind of don't know what i'm doing, I understand the values and what they do and that the more feature maps the more things it can learn so I just "threw in" some model I found [here](http://cs231n.github.io/convolutional-networks/). As for using pretrained models I have no clue where to begin.

When computing the images I don't actually change anything, the network and networktrainer is loaded with the same values and I don't ever touch them again.
I cleaned up my compute code and commented it using regions so you can see what it does.

In the InputLayer I extended the FeedData function with one that takes a float[] and made the imageChannels, imageHeight, and imageWidth public
![62157b93c5332796d8578ed742100567](https://cloud.githubusercontent.com/assets/3842800/16882965/f244490e-4ac1-11e6-9c7c-78e9ba962c5a.png)

I also added a Compute function to the NeuralNetwork.cs file
![80c4d60d2ade682e953ba577830c73ab](https://cloud.githubusercontent.com/assets/3842800/16882988/14c4ac44-4ac2-11e6-9abd-ac48242c8903.png)

I run the Compute function with an image(bitmap) twice which returns different results.

**Comment by jcredi** [comment]
That's the point, I'm not exactly a developer, and it was my first time with C# (as you can probably tell, e.g. from the fact that I didn't use Images or Bitmaps), so I'm really happy you figured out how to use my code. Actually, all your modifications make a lot of sense. Let me know if you wish to be added as a contributor to the repo.

Ok so since it doesn't crash when using your CPU, it looks like there's something in my OpenCL kernels for the maxpooling layer that the AMD APP compiler doesn't like very much. I hope I'll have some time in the weekend to go thoroughly through this issue.
Just out of curiosity, can you reproduce the non-crashing behaviour if using device 1, i.e. the Intel CPU, but with AMD APP platform?

The black boxes still remain a mystery to me. I tried googling for a bit, and it looks like it usually happens when the GPU is overheating (which doesn't seem to be the case) or faulty (hopefully not either!). Is this the first time this occurs? 
Another thing that just popped in my mind is memory leaking. It's quite easy to forget releasing all OpenCL events and GPU memory can be quickly saturated with junk when that happens. Now, I have no idea if that can be related to the artifacts on your display, but it may be worth checking GPU memory usage while you run the program.

No idea about the crash when moving an idle VS window, sorry! 

The Compute function looks good and you are (correctly) only carrying out the forward pass, so you _shouldn't_ really be getting different results when passing the same image twice... This is quite unsettling, I'll have to check it out tomorrow.

By the way, the Stanford course you linked is an excellent resource for getting familiar with CNNs. If you're interested in knowing more, that's one of the two places I'd really recommend starting from (the other one is [this free online book](http://neuralnetworksanddeeplearning.com/)).
They actually talk about using pretrained model in Stanford CS231n ([here](https://cs231n.github.io/transfer-learning/)). Depending on what you're trying to do, it may really be worth a look.

As an aside, if your data set size keeps increasing and training speed starts to be relevant, you may want to experiment with larger mini-batches when training (doesn't have to factor  That will allow your GPU to compute larger matrix multiplications less often, rather than small matrix multiplications more often (thus taking advantage of the 2048 shading units in your GPU ;) ). Once the net is trained, you can always save it to file, reload it, and set the mini-batch size to 1 for using your Compute function.

EDIT: forgot to say you can use any mini-batch size you like (i.e. your data set size does not need to divide the mini-batch size). Also, concerning performance, you may want to play with the BASE_GROUP_SIZE (architecture dependent) and OPTIMAL_GROUP_SIZE (device dependent) constants in OpenCLSpace.cs.

**Comment by merceyz** [comment]
I thought that was a little bit weird but it's not the first place I've encountered people not using images for image recognition but rather weird looking files. I would love to help develop on this project. Regarding communication we should probably find another way, Skype perhaps?

It still crashes without the maxpool layer but it's just way faster the few times it works
I executed it on device 2(1) and it instantly throws an exception so that counts as a crash

> Exception thrown at 0x000007FECE424DA3 (amdocl64.dll) in Console for testing.exe: 0xC0000005: Access violation reading location 0x00000000000000DC.
> 
> If there is a handler for this exception, the program may be safely continued.

The GPU is relatively new so it **shouldn't** be anything wrong with it. However while playing games I sometimes get "screen tearing" and horrible graphical bugs. When it crashes it is at about 300mb of ram, crash happens, it recovers and then it's at around 165

As for the crashing VS window that would be the screen card driver "giving up" after running the program

I have also read that one, even based one of my earlier models on it but sadly it didn't yield any good results (different library/framework)

I will make sure to increase those values before training, thanks!

When I played with those values it also crashed instantly so I think i'll leave those alone for now.

EDIT:
I edited the model a little bit and actually got it to run for a while (still running while typing this.) Also gpu at 447 MB so something is probably leaking.
![356a279b977364511dfa80c46c475412](https://cloud.githubusercontent.com/assets/3842800/16891597/a1ab0b32-4afc-11e6-90b8-e8e3cba2bb88.png)
I got it to an error value of 0
![e296b60ea9a47c158cd7b7c1659aa2e1](https://cloud.githubusercontent.com/assets/3842800/16891636/36b1c536-4afd-11e6-9b48-89d3bac9f762.png)

**Comment by merceyz** [comment]
I edited the code to add some more logging to catch where it throws the error, I got different layers, locations and functions at every crash so it doesn't help me pinpoint it much but it might help you.
Here are a few of them
- Loop starts in NetworkEvaluator.cs Line 67
  ![b26e761c002fb1d389cd5e4cecbdb3a5](https://cloud.githubusercontent.com/assets/3842800/16898238/318ede90-4bcf-11e6-984c-d6c5263a08ab.png)
  ![5c7a6d699d32b90ab0643feb5fdd2341](https://cloud.githubusercontent.com/assets/3842800/16898240/3631a39c-4bcf-11e6-9693-ef49cd60c409.png)
  In other words somewhere in the forward pass
  So I went to the forward pass function in NeuralNetwork.cs and made it write out the "l" value (current layer) it gets to 16, counting my layers (with the input layer as 0) that is the final FC layer before the SoftMax layer
  Tracked it down to this call that crashes it
  ![71407e4e5399ad3a650a8cb7076ee2a2](https://cloud.githubusercontent.com/assets/3842800/16898345/b579036a-4bd5-11e6-9bc6-382daaca59ad.png)
- Loop starts in NetworkTrainer.cs line 365
  ![ed898ab2da58fc5ab6e6541651ba5c24](https://cloud.githubusercontent.com/assets/3842800/16898270/30e0982e-4bd1-11e6-985a-764a95bce176.png)
  Results in display driver crashing when indexCounter hits 7 aka the backward pass
  Into the backwards pass function we go, it gets to to layer 4 which is a convlayer.
  It gets to the BackPropagate function in the conv layer, tracked down to this call
  ![e2d36de91d185e0ac714a2329f3bb86c](https://cloud.githubusercontent.com/assets/3842800/16898386/b0118228-4bd8-11e6-9be0-2c4442667d45.png)
  ![cb903aed10667a4fba5731b68c8d5be6](https://cloud.githubusercontent.com/assets/3842800/16898388/e828610e-4bd8-11e6-85b7-dfceb58c06b1.png)

**Comment by jcredi** [comment]
Great job! That's valuable info, and I'll have a lot to look at.

I was planning to install the AMD APP sdk and try to reproduce your errors on my Intel CPU. Unfortunately, though, my desktop suddenly stopped working yesterday morning and I'm still trying to figure out what is wrong. My current guess is that my motherboard is dead, and if I'm right I'm afraid you'll be on your own until I get a new one :(
(I also have a laptop, but it runs linux and I can't risk playing around with dual boots as I need it to work.)

It's great that you got to 0 error! However, remember that neural nets are in general very powerful models and it's relatively easy to overfit your training data. I know your data set is quite small, but you should really try to put some images aside and use them for validation and testing!

**Comment by merceyz** [comment]
I hope you manage to figure it out and that it isn't too expensive of a fix.

I'll make sure to take some out for validation and testing but do you suggest having the same amount of data per label (100 images for label 1, 100 for label 2 etc)?

**Comment by jcredi** [comment]
In general it's nice having (roughly) balanced class frequencies, but it is also very common to have one or few labels with very few examples. When that's the case, it's a good idea to keep the same label ratio in splitting your data into training, validation, and test set. If you have, say, 100 examples with label A and 20 with label B, you may create a validation set by randomly pick 20 images with label A and 4 with label B (20% in both cases).

**Comment by merceyz** [comment]
Any news relating to my issues?
